transcribers/mqtt.py

```diff
@@ -131,28 +133,53 @@ def _mqtt_to_ipal(self, request, pkt):
-        match_types = MQTTProtocol.command_response(response)
-        if len(match_types) == 0:
-            return []
-
-        # then only check relevant types
         for request in requests:
-            if request.type in match_types:
-                if (request.src, request.dest) == (response.dest, response.src):
-                    if request.type in {MQTTProtocol.CONNECT, MQTTProtocol.PINGREQ}:
-                        response.responds_to.append(request.id)
-                        remove_from_queue.append(request)
-
-                    elif (
-                        request.type
-                        in MQTTProtocol.TOPIC_COMMANDS | MQTTProtocol.PUBLISH_COMMANDS
-                        and request.keys() == response.keys()
-                    ):
-                        response.responds_to.append(request.id)
-                        remove_from_queue.append(request)
+            if response.type == MQTTProtocol.PUBACK:
+                if request.type == MQTTProtocol.PUBLISH and request.src == response.dest:
+                    response.responds_to.append(request.id)
+                    remove_from_queue.append(request)
+                    break
+
+            elif response.type == MQTTProtocol.PUBREC:  # PubREC
+                if request.type == MQTTProtocol.PUBLISH and request.src == response.dest:
+                    response.responds_to.append(request.id)
+                    # do not remove request from queue because PUBCOMP is part of the 4-way exchange MQTT with in QoS 2
+
+            elif response.type == MQTTProtocol.PUBCOMP:  # PubCOMP
+                # completes the 4 way exchange of QoS 2 MQTT
+                # includes two requests: PUBLISH and PUBREL
+                if request.type == MQTTProtocol.PUBLISH and request.src == response.dest:
+                    response.responds_to.append(request.id)
+                    remove_from_queue.append(request)
+                elif request.type == MQTTProtocol.PUBREL and request.src == response.dest:
+                    response.responds_to.append(request.id)
+                    remove_from_queue.append(request)
+
+            elif response.type == MQTTProtocol.SUBACK:
+                # check for matching topics
+                req_topics = list(request.data.keys())
+                res_topics = list(response.data.keys())
+                if set(res_topics).issubset(req_topics):
+                    print(set(res_topics), set(req_topics))
+                    response.responds_to.append(request.id)
+                    remove_from_queue.append(request)
+
+            # all other message types are matched by source and destination and by the following message types:
+            # Connect 1 -> ConnACK 2
+            # Unsubscribe 10 -> UnsubACK 11
+            # PingREQ 12 -> PingRESP 13
+            elif request.type in [MQTTProtocol.CONNECT, MQTTProtocol.UNSUBSCRIBE, MQTTProtocol.PINGREQ] and \
```

Comment:
This list should be a class member accessed via MQTTProtocol.SIMPLE_REQUESTS

Reply:
I have added this in the latest commit